libs/react-spa/bff/src/lib/BffProvider.tsx

```diff
@@ -0,0 +1,240 @@
+  /**
+   * The base path of the application.
+   */
+  applicationBasePath: string
+}
+
+export const BffProvider = ({
+  children,
+  applicationBasePath,
+}: BffProviderProps) => {
+  const [showSessionExpiredScreen, setSessionExpiredScreen] = useState(false)
+  const bffUrlGenerator = createBffUrlGenerator(applicationBasePath)
+  const [state, dispatch] = useReducer(reducer, initialState)
+
+  const { authState } = state
+  const showErrorScreen = authState === 'error'
+  const showLoadingScreen =
+    authState === 'loading' ||
+    authState === 'switching' ||
+    authState === 'logging-out'
+  const isLoggedIn = authState === 'logged-in'
+
+  const { postMessage } = useBffBroadcaster((event) => {
+    if (
+      isLoggedIn &&
+      event.data.type === BffBroadcastEvents.NEW_SESSION &&
+      isNewSession(state.userInfo, event.data.userInfo)
+    ) {
+      setSessionExpiredScreen(true)
+    } else if (event.data.type === BffBroadcastEvents.LOGOUT) {
+      dispatch({
+        type: ActionType.LOGGED_OUT,
+      })
+
+      signIn()
+    }
+  })
+
+  useEffect(() => {
+    if (isLoggedIn) {
+      // Broadcast to all tabs/windows/iframes that a new session has started
+      postMessage({
+        type: BffBroadcastEvents.NEW_SESSION,
+        userInfo: state.userInfo,
+      })
+    }
+  }, [postMessage, state.userInfo, isLoggedIn])
+
+  const checkLogin = async (noRefresh = false) => {
+    dispatch({
+      type: ActionType.SIGNIN_START,
+    })
+
+    try {
+      const url = bffUrlGenerator('/user', {
+        no_refresh: noRefresh.toString(),
+      })
+
+      const res = await fetch(url, {
+        credentials: 'include',
+      })
+
+      if (!res.ok) {
+        // Bff server is down
+        if (res.status >= 500) {
+          throw new Error(BFF_SERVER_UNAVAILABLE)
+        }
+
+        // For other none ok responses, like 401/403, proceed with sign-in redirect.
+        signIn()
+
+        return
+      }
+
+      const user = await res.json()
+
+      dispatch({
+        type: ActionType.SIGNIN_SUCCESS,
+        payload: user,
+      })
+    } catch (error) {
+      dispatch({
+        type: ActionType.ERROR,
+        payload: error,
+      })
+    }
+  }
+
+  const signIn = useCallback(() => {
+    dispatch({
+      type: ActionType.SIGNIN_START,
+    })
+
+    window.location.href = bffUrlGenerator('/login', {
+      target_link_uri: window.location.href,
+    })
+  }, [bffUrlGenerator])
+
+  const signOut = useCallback(() => {
+    if (!state.userInfo) {
+      return
+    }
+
+    dispatch({
+      type: ActionType.LOGGING_OUT,
+    })
+
+    window.location.href = bffUrlGenerator('/logout', {
+      sid: state.userInfo.profile.sid,
+    })
+
+    setTimeout(() => {
+      // We will wait 5 seconds before we post the logout message to other tabs/windows/iframes.
+      // The reason is that IDS will not log the user out immediately.
+      postMessage({
+        type: BffBroadcastEvents.LOGOUT,
+      })
+    }, 5000)
```

Comment:
This timeout probably never runs, because the window is unloaded while redirecting to bff/logout.

Need to post the logout before redirecting, and instead wait a few seconds on the other side. Though the bff poller may have triggered logout by that time anyways.